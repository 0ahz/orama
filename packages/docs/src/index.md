---
home: true
heroImage: https://nearform.github.io/lyra/imgs/lyra.svg
tagline: Documentation for the Lyra search engine
actionText: Quick Start →
actionLink: /guide/
features:
- title: Fast
  details: Lyra lookups takes microseconds.
- title: In-memory
  details: Lyra keeps data and indices in-memory.
- title: Portable
  details: Can work both on client and server.
footer: Made by NearForm with ❤️
---
