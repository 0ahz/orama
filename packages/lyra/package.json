{
  "name": "@nearform/lyra",
  "description": "Fast, in-memory, full-text search engine",
  "version": "0.0.1",
  "author": {
    "name": "Michele Riva",
    "email": "ciao@micheleriva.it",
    "url": "https://github.com/MicheleRiva"
  },
  "bugs": {
    "url": "https://github.com/nearform/lyra"
  },
  "scripts": {
    "build:umd": "swc **/*.ts --ignore **/*.test.ts -d ./dist/umd -s -C module.type=umd",
    "build:module": "swc **/*.ts --ignore **/*.test.ts -d ./dist/esm -s -C module.type=es6",
    "build:types": "tsc **/*.ts --declaration --emitDeclarationOnly --outdir ./dist/types",
    "build": "pnpm build:umd && pnpm build:module && pnpm build:types"
  },
  "main": "./dist/umd/lyra.js",
  "module": "./dist/esm/lyra.js",
  "types": "./dist/types/src/lyra.d.ts",
  "dependencies": {
    "fastq": "^1.13.0",
    "nanoid": "3.3.4"
  },
  "devDependencies": {
    "@swc/cli": "^0.1.57",
    "@swc/core": "^1.2.185",
    "@types/jest": "^27.5.0",
    "jest": "^28.1.0",
    "ts-jest": "^28.0.2",
    "typescript": "^4.6.4"
  }
}