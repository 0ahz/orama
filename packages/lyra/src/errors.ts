function formatJSON(input: object) {
  return JSON.stringify(input, null, 2);
}

export const INVALID_SCHEMA_TYPE = (type: string) =>
  `Invalid schema type. Expected string or object, but got ${type}`;
export const INVALID_DOC_SCHEMA = (expected: object, found: object) =>
  `Invalid document structure. \nClass has been initialized with the following schema: \n\n${formatJSON(
    expected
  )}\n\nbut found the following doc:\n\n${formatJSON(found)}`;
