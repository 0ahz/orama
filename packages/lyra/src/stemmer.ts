import natural from "natural";

export type Language = typeof SUPPORTED_LANGUAGES[number];

export const SUPPORTED_LANGUAGES = [
  "dutch",
  "english",
  "french",
  "italian",
  "norwegian",
  "portugese",
  "russian",
  "spanish",
  "swedish",
] as const;

export function stemArray(input: string[], language: Language): string[] {
  let stemmer;

  switch (language) {
    case "dutch":
      stemmer = natural.PorterStemmerNl;
      break;
    case "english":
      stemmer = natural.PorterStemmer;
      break;
    case "french":
      stemmer = natural.PorterStemmerFr;
      break;
    case "italian":
      stemmer = natural.PorterStemmerIt;
      break;
    case "norwegian":
      stemmer = natural.PorterStemmerNo;
      break;
    case "portugese":
      stemmer = natural.PorterStemmerPt;
      break;
    case "russian":
      stemmer = natural.PorterStemmerRu;
      break;
    case "spanish":
      stemmer = natural.PorterStemmerEs;
      break;
    case "swedish":
      stemmer = natural.PorterStemmerSv;
      break;
    default:
      break;
  }

  if (stemmer) {
    return input.map(stemmer.stem);
  }

  return input;
}
