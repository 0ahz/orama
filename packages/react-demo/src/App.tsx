import { useEffect, useState } from 'react';
import { Lyra } from '@nearform/lyra/dist/esm/lyra';

const db = new Lyra({
  schema: {
    id: 'number',
    name: 'string',
  }
});

function App() {
  const [isLoading, setIsLoading] = useState(true);
  const [pokemon, setPokemon] = useState<any[]>();
  const [searchPokemon, setSearchPokemon] = useState<any[]>();
  const [search, setSearch] = useState('');

  useEffect(() => {
    fetch('/pokedex.json')
      .then((data) => data.json())
      .then(async (pokeList) => {
        for (const { id, name } of pokeList.pokemon) {
          await db.insert({
            id, name
          });
        }

        setPokemon(pokeList.pokemon);
      })
      .catch(console.log)
      .finally(() => setIsLoading(false));
  }, []);

  useEffect(() => {
    if (search) {
      db.search({
        term: search,
        properties: '*'
      })
        .then((result) => {
          const pokemons = result.hits.map((d) => pokemon?.find(p => p.id === (d as any).id));

          if (pokemons.length && !pokemons.some((x) => !x)) {
            setSearchPokemon([...new Set(pokemons)]);
          } else {
            setSearchPokemon([]);
          }
        })
        .catch(console.log)
    }
  }, [search]);

  return (
    <div>
      <h1> Lyra preview in React. Search for a Pokemon </h1>

      {isLoading && (
        <div> Loading pokemons... </div>
      )}

      <div>
        <input type='text' value={search} onChange={(e) => setSearch(e.target.value)} />
      </div>

      {!isLoading && (
        <div className='grid'>
          {searchPokemon?.map((p) => (
            <div className='grid grid-cols-2' key={p.id}>
              <img src={p.img} />
              <div>
                <h1 className='text-lg'>
                  {p.name}
                </h1>
                <p> Type: {p.type.join(', ')} </p>
              </div>
            </div>
          ))}
        </div>
      )}
    </div>
  )
}

export default App
