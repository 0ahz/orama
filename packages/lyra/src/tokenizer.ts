import { WordTokenizer } from "natural";
import { Language, stemArray } from "./stemmer";

const tokenizer = new WordTokenizer();

export function tokenize(
  input: string,
  language: Language = "english"
): Set<string> {
  const tokens = tokenizer.tokenize(input);
  return new Set(stemArray(tokens, language));
}
