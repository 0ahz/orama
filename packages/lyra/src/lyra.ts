import type { queueAsPromised } from "fastq";
import { nanoid } from "nanoid";
import * as fastq from "fastq";
import { Trie } from "./prefix-tree/trie";
import * as ERRORS from "./errors";
import { tokenize } from "./tokenizer";
import { formatNanoseconds, getNanosecondsTime } from "./utils";

export type PropertyType = "string" | "number" | "boolean";

export type PropertiesSchema = {
  [key: string]: PropertyType | PropertiesSchema;
};

export type LyraProperties = {
  schema: PropertiesSchema;
};

export type LyraDocs = Map<string, object>;

export type SearchParams = {
  term: string;
  properties?: "*" | string[];
  limit?: number;
  offset?: number;
};

type LyraIndex = Map<string, Trie>;

type QueueDocParams = {
  id: string;
  doc: object;
};

type SearchResult = Promise<{
  count: number;
  hits: object[];
  elapsed: string;
}>;

export class Lyra {
  private schema: PropertiesSchema;
  private docs: LyraDocs = new Map();
  private index: LyraIndex = new Map();
  private queue: queueAsPromised<QueueDocParams> = fastq.promise(
    this,
    this._insert,
    1
  );

  constructor(properties: LyraProperties) {
    this.schema = properties.schema;
    this.buildIndex(properties.schema);
  }

  private buildIndex(schema: PropertiesSchema) {
    for (const prop in schema) {
      const propType = typeof prop;

      if (propType === "string") {
        this.index.set(prop, new Trie());
      } else if (propType === "object") {
        this.buildIndex(prop as unknown as PropertiesSchema);
      } else {
        throw ERRORS.INVALID_SCHEMA_TYPE(propType);
      }
    }
  }

  async search(params: SearchParams): SearchResult {
    const tokens = tokenize(params.term).values();
    const indices = this.getIndices(params.properties);
    const results: object[] = [];

    const timeStart = getNanosecondsTime();

    for (const token of tokens) {
      for (const index of indices) {
        const searchResult = await this._search({
          ...params,
          index: index,
          term: token,
        });

        if (searchResult.length) {
          for (const res of searchResult) {
            results.push(res);
          }
        }
      }
    }

    return {
      elapsed: formatNanoseconds(getNanosecondsTime() - timeStart),
      hits: results.slice(params.offset ?? 0, params.limit ?? 10), // @todo avoid getting all results and slicing them
      count: results.length,
    };
  }

  private getIndices(indices: SearchParams["properties"]): string[] {
    const knownIndices = [...this.index.keys()];

    if (!indices) {
      return knownIndices;
    }

    if (typeof indices === "string") {
      if (indices === "*") {
        return knownIndices;
      } else {
        throw ERRORS.INVALID_PROPERTY(indices, knownIndices);
      }
    }

    for (const index of indices as string[]) {
      if (!knownIndices.includes(index)) {
        throw ERRORS.INVALID_PROPERTY(index, knownIndices);
      }
    }

    return indices as string[];
  }

  private async _search(
    params: SearchParams & { index: string }
  ): Promise<object[]> {
    const idx = this.index.get(params.index);
    const searchResult = idx?.find(params.term);
    const results: object[] = [];

    for (const key in searchResult) {
      const docs: string[] = [];

      for (const id of (searchResult as any)[key]) {
        const fullDoc = this.docs.get(id);
        // eslint-disable-next-line @typescript-eslint/ban-ts-comment
        // @ts-ignore
        docs.push({ id, ...fullDoc });
      }

      results.push(docs);
    }

    return results.flat();
  }

  public async insert(doc: object): Promise<{ id: string }> {
    const id = nanoid();

    if (!(await this.checkInsertDocSchema(doc))) {
      throw ERRORS.INVALID_DOC_SCHEMA(this.schema, doc);
    }

    await this.queue.push({
      id,
      doc,
    });

    return { id };
  }

  private async _insert({ doc, id }: QueueDocParams): Promise<void> {
    const index = this.index;
    this.docs.set(id, doc);

    function recursiveTrieInsertion(doc: object) {
      for (const key in doc) {
        if (typeof (doc as any)[key] === "object") {
          recursiveTrieInsertion((doc as any)[key]);
        } else {
          const requestedTrie = index.get(key);
          const tokens = tokenize((doc as any)[key]);

          for (const token of tokens) {
            requestedTrie?.insert(token, id);
          }
        }
      }
    }

    recursiveTrieInsertion(doc);
  }

  private async checkInsertDocSchema(
    doc: QueueDocParams["doc"]
  ): Promise<boolean> {
    function recursiveCheck(
      newDoc: QueueDocParams["doc"],
      schema: PropertiesSchema
    ): boolean {
      for (const key in newDoc) {
        if (!(key in schema)) {
          return false;
        }

        const propType = typeof (newDoc as any)[key];

        if (propType === "object") {
          recursiveCheck((newDoc as any)[key], schema);
        } else {
          if (typeof (newDoc as any)[key] !== schema[key]) {
            return false;
          }
        }
      }

      return true;
    }

    return recursiveCheck(doc, this.schema);
  }
}
