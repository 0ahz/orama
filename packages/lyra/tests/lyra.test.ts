import { Lyra } from "../src/lyra";

function getId({ id }: { id: string }) {
  return id;
}

describe("checkInsertDocSchema", () => {
  it("should compare the inserted doc with the schema definition", async () => {
    const db = new Lyra({
      schema: {
        quote: "string",
        author: "string",
      },
    });

    expect(
      await db.insert({ quote: "hello, world!", author: "me" }).then(getId)
    ).toBeDefined();

    try {
      await db.insert({ quote: "hello, world!", author: true });
    } catch (err) {
      expect(err).toMatchSnapshot();
    }

    try {
      await db.insert({
        quote: "hello, world!",
        authors: "author should be singular",
      });
    } catch (err) {
      expect(err).toMatchSnapshot();
    }

    try {
      await db.insert({ quote: "hello, world!", foo: { bar: 10 } });
    } catch (err) {
      expect(err).toMatchSnapshot();
    }
  });
});

describe("lyra", () => {
  it("should correctly insert and retrieve data", async () => {
    const db = new Lyra({
      schema: {
        example: "string",
      },
    });

    const ex1Insert = await db.insert({ example: "The quick, brown, fox" });
    const ex1Search = await db.search({
      term: "quick",
      properties: ["example"],
    });

    expect(ex1Insert.id).toBeDefined();
    expect(ex1Search.count).toBe(1);
    expect(ex1Search.elapsed).toBeDefined();
    expect((ex1Search.hits[0] as any).example).toBe("The quick, brown, fox");
  });

  it("should correctly paginate results", async () => {
    const db = new Lyra({
      schema: {
        animal: "string",
      },
    });

    await db.insert({ animal: "Quick brown fox" });
    await db.insert({ animal: "Lazy dog" });
    await db.insert({ animal: "Jumping penguin" });
    await db.insert({ animal: "Fast chicken" });
    await db.insert({ animal: "Fabolous ducks" });
    await db.insert({ animal: "Fantastic horse" });

    const search1 = await db.search({ term: "f", limit: 1, offset: 0 });
    const search2 = await db.search({ term: "f", limit: 1, offset: 1 });
    const search3 = await db.search({ term: "f", limit: 1, offset: 2 });
    const search4 = await db.search({ term: "f", limit: 2, offset: 2 });

    expect(search1.count).toBe(4);
    expect((search1.hits[0] as any).animal).toBe("Quick brown fox");

    expect(search2.count).toBe(4);
    expect((search2.hits[0] as any).animal).toBe("Fast chicken");

    expect(search3.count).toBe(4);
    expect((search3.hits[0] as any).animal).toBe("Fabolous ducks");

    expect(search4.count).toBe(4);
    expect((search4.hits[0] as any).animal).toBe("Fabolous ducks");
    expect((search4.hits[1] as any).animal).toBe("Fantastic horse");
  });

  it("Should throw an error when searching in non-existing indices", async () => {
    const db = new Lyra({ schema: { foo: "string", baz: "string" } });

    try {
      await db.search({
        term: "foo",
        properties: ["bar"],
      });
    } catch (err) {
      expect(err).toMatchInlineSnapshot(
        `"Invalid property name. Expected a wildcard string (\\"*\\") or array containing one of the following properties: foo, baz, but got: bar"`
      );
    }
  });

  it("Should correctly remove a document after its insertion", async () => {
    const db = new Lyra({
      schema: {
        quote: "string",
        author: "string",
      },
    });

    const { id: id1 } = await db.insert({
      quote: "Be yourself; everyone else is already taken.",
      author: "Oscar Wilde",
    });

    const { id: id2 } = await db.insert({
      quote:
        "To live is the rarest thing in the world. Most people exist, that is all.",
      author: "Oscar Wilde",
    });

    await db.insert({
      quote: "So many books, so little time.",
      author: "Frank Zappa",
    });

    const res = await db.delete(id1);

    const searchResult = await db.search({
      term: "Oscar",
      properties: ["author"],
    });

    expect(res).toBeTruthy();
    expect(searchResult.count).toBe(1);
    expect((searchResult.hits[0] as any).author).toBe("Oscar Wilde");
    expect((searchResult.hits[0] as any).quote).toBe(
      "To live is the rarest thing in the world. Most people exist, that is all."
    );
    expect((searchResult.hits[0] as any).id).toBe(id2);
  });

  it("Should preserve identical docs after deletion", async () => {
    const db = new Lyra({
      schema: {
        quote: "string",
        author: "string",
      },
    });

    const { id: id1 } = await db.insert({
      quote: "Be yourself; everyone else is already taken.",
      author: "Oscar Wilde",
    });

    const { id: id2 } = await db.insert({
      quote: "Be yourself; everyone else is already taken.",
      author: "Oscar Wilde",
    });

    await db.insert({
      quote: "So many books, so little time.",
      author: "Frank Zappa",
    });

    const res = await db.delete(id1);

    const searchResult = await db.search({
      term: "Oscar",
      properties: ["author"],
    });

    const searchResult2 = await db.search({
      term: "already",
      properties: ["quote"],
    });

    expect(res).toBeTruthy();
    expect(searchResult.count).toBe(1);
    expect((searchResult.hits[0] as any).author).toBe("Oscar Wilde");
    expect((searchResult.hits[0] as any).quote).toBe(
      "Be yourself; everyone else is already taken."
    );
    expect((searchResult.hits[0] as any).id).toBe(id2);

    expect(searchResult2.count).toBe(1);
    expect((searchResult2.hits[0] as any).author).toBe("Oscar Wilde");
    expect((searchResult2.hits[0] as any).quote).toBe(
      "Be yourself; everyone else is already taken."
    );
    expect((searchResult2.hits[0] as any).id).toBe(id2);
  });
});
