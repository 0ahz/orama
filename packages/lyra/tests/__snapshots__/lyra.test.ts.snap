// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`checkInsertDocSchema should compare the inserted doc with the schema definition 1`] = `
"Invalid document structure. 
Class has been initialized with the following schema: 

{
  \\"quote\\": \\"string\\",
  \\"author\\": \\"string\\"
}

but found the following doc:

{
  \\"quote\\": \\"hello, world!\\",
  \\"author\\": true
}"
`;

exports[`checkInsertDocSchema should compare the inserted doc with the schema definition 2`] = `
"Invalid document structure. 
Class has been initialized with the following schema: 

{
  \\"quote\\": \\"string\\",
  \\"author\\": \\"string\\"
}

but found the following doc:

{
  \\"quote\\": \\"hello, world!\\",
  \\"authors\\": \\"author should be singular\\"
}"
`;

exports[`checkInsertDocSchema should compare the inserted doc with the schema definition 3`] = `
"Invalid document structure. 
Class has been initialized with the following schema: 

{
  \\"quote\\": \\"string\\",
  \\"author\\": \\"string\\"
}

but found the following doc:

{
  \\"quote\\": \\"hello, world!\\",
  \\"foo\\": {
    \\"bar\\": 10
  }
}"
`;
